### PR TITLE
Use formulae to parse model formulas

### PR DESCRIPTION
**Pull request by tomicapretto**
This is quite a big PR where I modified many things in how Bambi works to incorporate [formulae](https://github.com/bambinos/formulae) into its usage. Main changes are listed below

**New features**

1. Simplified syntax to specify mixed models
```
model = Model(data)
fitted = model.fit("y ~ x + (x|g)")
```

2. New alias for `I()`

```
fitted1 = model.fit("y ~ {x + z}")
fitted2 = model.fit("y ~ I(x + z)")
```
`fitted1` and `fitted2` are equivalent. 

3. We can use non-syntactic names

```
fitted = model.fit("y ~ `My variable with spaces` + `$!W31rd N4m3##`")
```

Those are valid as dataframe column names but invalid as variable names in Python. Patsy implementation doesn't let you use those variables, but formulae does. 

Ideas 2 and 3 are taken from  [formulaic](https://github.com/matthewwardrop/formulaic). It is a great package but it does not support mixed effects.

**Dropped features**

1. You can't do `model.add()` anymore. 
2. All Patsy specific encodings for categorical variables. formulae uses treatment encoding by default. Other encodings are planned to be implemented in the future.

There may be more changes that I'm not remembering right now. Please participate in the discussion if you have anything to say,   I would like to have as many opinions as possible about these changes.

### CODE REVIEW

File: bambi/models.py [hunk 1]
```diff
@@ -7,17 +7,18 @@
...
+
 from arviz.plots import plot_posterior
 from arviz.data import from_dict
-from patsy import dmatrices, dmatrix, EvalFactor
-import pymc3 as pm
+from formulae import design_matrices
 
-import bambi.version as version
 from .backends import PyMC3BackEnd
-from .external.patsy import Custom_NA
```
Comment (aloctavodia) [inline bambi/models.py:18]: I guess we can delete this module

File: bambi/models.py [hunk 2]
```diff
@@ -1018,171 +739,3 @@ def common_terms(self):
     def group_specific_terms(self):
         """Return dict of all and only group specific effects in model."""
         return {k: v for (k, v) in self.terms.items() if v.group_specific}
-
-
-class BaseTerm:
```
Comment (aloctavodia) [inline bambi/models.py:1023]: Nice you move all these

**Comment by aloctavodia** [comment]
Right, quite big! But test are passing haha. I will try to give another read tomorrow. Overall the impression is that code is much cleaner now.

**Comment by twiecki** [comment]
image

This is great stuff!

File: bambi/tests/test_built_models.py [hunk 3]
```diff
@@ -36,38 +36,25 @@ def crossed_data():
 
 def test_empty_model(crossed_data):
```
Comment (aloctavodia) [inline bambi/tests/test_built_models.py:37]: This used to test that two different ways to create models were actually equivalent, but now it is just creating the same model in the same way twice.

File: bambi/tests/test_built_models.py [hunk 4]
```diff
@@ -36,38 +36,25 @@ def crossed_data():
...
-    model1.fit(tune=0, draws=1)
-
-    # check that both models have same priors for common effects
+    # check that both models have same priors for common effects -> emtpy priors
     priors0 = {x.name: x.prior.args for x in model0.terms.values() if not x.group_specific}
     priors1 = {x.name: x.prior.args for x in model1.terms.values() if not x.group_specific}
     assert set(priors0) == set(priors1)
 
 
 def test_intercept_only_model(crossed_data):
```
Comment (aloctavodia) [inline bambi/tests/test_built_models.py:50]: same as previous comment

File: bambi/tests/test_built_models.py [hunk 5]
```diff
@@ -76,16 +63,11 @@ def test_intercept_only_model(crossed_data):
 def test_slope_only_model(crossed_data):
```
Comment (aloctavodia) [inline bambi/tests/test_built_models.py:63]: same as previous comment

File: bambi/tests/test_built_models.py [hunk 6]
```diff
@@ -100,16 +82,11 @@ def test_slope_only_model(crossed_data):
 def test_cell_means_parameterization(crossed_data):
```
Comment (aloctavodia) [inline bambi/tests/test_built_models.py:82]: same as previous comment

File: bambi/tests/test_built_models.py [hunk 7]
```diff
@@ -140,35 +116,27 @@ def test_3x4_common_anova(crossed_data):
...
+    # without intercept (i.e., 2-factor cell means model)
     model1 = Model(crossed_data)
-    model1.fit("Y ~ 0 + threecats*fourcats", run=False)
-    model1.build(backend="pymc3")
-    fitted1 = model1.fit(tune=0, draws=1)
+    fitted1 = model1.fit("Y ~ 0 + threecats*fourcats", tune=0, draws=1)
     assert len(fitted1.posterior.data_vars) == 4
 
 
 def test_cell_means_with_covariate(crossed_data):
```
Comment (aloctavodia) [inline bambi/tests/test_built_models.py:130]: same as previous comment

File: bambi/tests/test_built_models.py [hunk 8]
```diff
@@ -207,36 +174,21 @@ def test_many_common_many_group_specific(crossed_data):
...
-    model1.add("dummy")
-    model1.add("threecats")
-    model1.add(group_specific="0+threecats|subj")
-    model1.add(group_specific="1|item")
-    model1.add(group_specific="0+continuous|item")
-    model1.add(group_specific="dummy|item")
-    model1.add(group_specific="threecats|site")
-    model1.build(backend="pymc3")
+    model1.fit(
+        "Y ~ continuous + dummy + threecats + (threecats|subj) + (continuous|item) + (dummy|item) + (threecats|site)",
```
Comment (aloctavodia) [inline bambi/tests/test_built_models.py:187]: The syntax is not the same as with model0, is this intended?

File: bambi/tests/test_built_models.py [hunk 9]
```diff
@@ -478,52 +423,38 @@ def dicts_close(a, b):
...
 
 def test_logistic_regression_categoric():
     y = pd.Series(np.random.choice(["a", "b"], 50), dtype="category")
     data = pd.DataFrame({"y": y, "x": np.random.normal(size=50)})
     model = Model(data)
-    fitted = model.fit("y ~ x", family="bernoulli")
+    model.fit("y ~ x", family="bernoulli")
 
 
 def test_poisson_regression(crossed_data):
```
Comment (aloctavodia) [inline bambi/tests/test_built_models.py:449]: This is defining the same model twice

File: requirements.txt [hunk 10]
```diff
@@ -4,3 +4,4 @@ pandas>=1.0.0
 patsy>=0.5.1
```
[the remark this answers is not in the document]
Comment (tomicapretto) [reply requirements.txt:4]: Nope

File: bambi/tests/test_built_models.py (hunk 8 above)
Comment (tomicapretto) [reply bambi/tests/test_built_models.py:187]: Yes. I wanted to see that implicit and explicit group specific intercepts work the same way.

File: bambi/tests/test_built_models.py (hunk 3 above)
Comment (tomicapretto) [reply bambi/tests/test_built_models.py:37]: You are right, and I was aware of that before the PR. I just didn't know if it was good to delete all the tests with this pattern. I decided to leave them just to see if fitting the same model twice, gives the same result both times... is it too dumb? do you suggest deleting those tests?

Comment (aloctavodia) [reply bambi/tests/test_built_models.py:37]: Remove them or just test that models build once.

Comment (tomicapretto) [reply bambi/tests/test_built_models.py:37]: Second option sounds better

**Review by aloctavodia — APPROVED** [review]
LGTM!

**Comment by tomicapretto** [comment]
Some notes for the record. 
We are not done with this PR. Many things regarding documentation (e.g. [Getting started](https://bambinos.github.io/bambi/getting_started.html)) have been written using the current version of Bambi (e.g. you can do `Model.add()`). We should start working in another PR to update docs after we merge this PR so the docs are in sync with how Bambi will work.

**Comment by aloctavodia** [comment]
The last commit seems a little bit unrelated o this PR.. I vote for merging this PRs ASAP and keep adding changes in future PRs.

**Comment by tomicapretto** [comment]
> The last commit seems a little bit unrelated o this PR.. I vote for merging this PRs ASAP and keep adding changes in future PRs.

I've just fixed one bug in formulae that is important in my opinion and now I'm re-running the notebook that made my discover the bug.

**PS:** The bug is related to index notation in response terms. Since the level of a variable can be anything that can go in a string, now you have to do `"response['this is my level!!'] ~ x + y + z"`, previously you could omit the quotes, but now you can't.

**PS2** Maybe tomorrow morning/evening this can be merged.

**Comment by tomicapretto** [comment]
I think this can be merged now

**Comment by tyarkoni** [comment]
I don't really have any useful comments here, just wanted to stick my head in to thank you for the lovely work—so, thanks!

**Comment by zwelitunyiswa** [comment]
I want to second what tyarkoni said. Thank you so much for making these tools accessible for non-technical people like me.

**Pull request merged**